File: lib/native/skill.ts
/**
 * Skill Native Module - Tauri commands for skill management
 *
 * Provides access to the Rust-based skill service for:
 * - Repository management (list, add, remove, toggle)
 * - Skill discovery from remote repositories
 * - Skill installation and uninstallation
 * - Skill state management (enable, disable, update)
 * - Skill content reading
 */

import { invoke } from '@tauri-apps/api/core';
import { isTauri } from '@/lib/native/utils';

// ========== Types ==========

/**
 * Skill repository configuration
 */
export interface SkillRepo {
  owner: string;
  name: string;
  branch: string;
  enabled: boolean;
}

/**
 * Discoverable skill from a repository
 */
export interface DiscoverableSkill {
  key: string;
  name: string;
  description: string;
  directory: string;
  readmeUrl: string | null;
  repoOwner: string;
  repoName: string;
  repoBranch: string;
}

/**
 * Installed skill record
 */
export interface InstalledSkill {
  id: string;
  name: string;
  description: string | null;
  directory: string;
  repoOwner: string | null;
  repoName: string | null;
  repoBranch: string | null;
  readmeUrl: string | null;
  installedAt: number;
  enabled: boolean;
  category: string | null;
  tags: string[];
}

/**
 * Combined skill view
 */
export interface NativeSkill {
  key: string;
  name: string;
  description: string;
  directory: string;
  readmeUrl: string | null;
  installed: boolean;
  enabled: boolean | null;
  repoOwner: string | null;
  repoName: string | null;
  repoBranch: string | null;
  category: string | null;
  tags: string[] | null;
}

/**
 * Local unregistered skill
 */
export interface LocalSkill {
  directory: string;
  name: string;
  description: string | null;
  path: string;
  hasSkillMd: boolean;
}

// ========== Repository Commands ==========

/**
 * List all configured skill repositories
 */
export async function listSkillRepos(): Promise<SkillRepo[]> {
  return invoke<SkillRepo[]>('skill_list_repos');
}

/**
 * Add a skill repository
 */
export async function addSkillRepo(
  owner: string,
  name: string,
  branch?: string
): Promise<void> {
  return invoke('skill_add_repo', { owner, name, branch });
}

/**
 * Remove a skill repository
 */
export async function removeSkillRepo(owner: string, name: string): Promise<void> {
  return invoke('skill_remove_repo', { owner, name });
}

/**
 * Toggle repository enabled state
 */
export async function toggleSkillRepo(
  owner: string,
  name: string,
  enabled: boolean
): Promise<void> {
  return invoke('skill_toggle_repo', { owner, name, enabled });
}

// ========== Discovery Commands ==========

/**
 * Discover skills from all enabled repositories
 */
export async function discoverSkills(): Promise<DiscoverableSkill[]> {
  return invoke<DiscoverableSkill[]>('skill_discover');
}

/**
 * Get all skills (merged view of discoverable, installed, and local)
 */
export async function getAllSkills(): Promise<NativeSkill[]> {
  return invoke<NativeSkill[]>('skill_get_all');
}

/**
 * Discover all skills (discoverable + installed + local)
 */
export async function discoverAllSkills(): Promise<{
  discoverable: DiscoverableSkill[];
  installed: InstalledSkill[];
  local: LocalSkill[];
}> {
  return invoke('skill_discover_all');
}

/**
 * Search and filter skills
 */
export async function searchSkills(filters: {
  category?: string;
  tags?: string[];
  installed?: boolean;
  enabled?: boolean;
  query?: string;
}): Promise<NativeSkill[]> {
  return invoke('skill_search', { filters });
}

/**
 * Scan for local unregistered skills
 */
export async function scanLocalSkills(): Promise<LocalSkill[]> {
  return invoke<LocalSkill[]>('skill_scan_local');
}

// ========== Installation Commands ==========

/**
 * Install a skill from repository
 */
export async function installSkill(skill: DiscoverableSkill): Promise<InstalledSkill> {
  return invoke<InstalledSkill>('skill_install', {
    owner: skill.repoOwner,
    repo: skill.repoName,
    branch: skill.repoBranch,
    directory: skill.directory,
    name: skill.name,
    description: skill.description,
    readmeUrl: skill.readmeUrl,
  });
}

/**
 * Install a skill from local path
 */
export async function installLocalSkill(
  sourcePath: string,
  name?: string
): Promise<InstalledSkill> {
  return invoke<InstalledSkill>('skill_install_local', { sourcePath, name });
}

/**
 * Register a local skill already in SSOT directory
 */
export async function registerLocalSkill(directory: string): Promise<InstalledSkill> {
  return invoke<InstalledSkill>('skill_register_local', { directory });
}

/**
 * Uninstall a skill
 */
export async function uninstallSkill(id: string): Promise<void> {
  return invoke('skill_uninstall', { id });
}

// ========== State Management Commands ==========

/**
 * Get all installed skills
 */
export async function getInstalledSkills(): Promise<InstalledSkill[]> {
  return invoke<InstalledSkill[]>('skill_get_installed');
}

export async function getSkill(id: string): Promise<InstalledSkill | null> {
  return invoke<InstalledSkill | null>('skill_get', { id });
}

/**
 * Get a specific installed skill or throw error
 */
export async function getSkillRequired(id: string): Promise<InstalledSkill> {
  return invoke<InstalledSkill>('skill_get_required', { id });
}

/**
 * Get skill state (legacy compatibility)
 */
export async function getSkillState(id: string): Promise<{
  installed: boolean;
  installedAt: string;
}> {
  return invoke('skill_get_state', { id });
}

/**
 * Enable a skill
 */
export async function enableSkill(id: string): Promise<void> {
  return invoke('skill_enable', { id });
}

/**
 * Disable a skill
 */
export async function disableSkill(id: string): Promise<void> {
  return invoke('skill_disable', { id });
}

/**
 * Update skill metadata
 */
export async function updateSkill(
  id: string,
  category?: string,
  tags?: string[]
): Promise<void> {
  return invoke('skill_update', { id, category, tags });
}

// ========== Content Commands ==========

/**
 * Read skill content (SKILL.md)
 */
export async function readSkillContent(directory: string): Promise<string> {
  return invoke<string>('skill_read_content', { directory });
}

/**
 * List skill resource files
 */
export async function listSkillResources(directory: string): Promise<string[]> {
  return invoke<string[]>('skill_list_resources', { directory });
}

/**
 * Read a skill resource file
 */
export async function readSkillResource(
  directory: string,
  resourcePath: string
): Promise<string> {
  return invoke<string>('skill_read_resource', { directory, resourcePath });
}

/**
 * Get SSOT directory path
 */
export async function getSkillSsotDir(): Promise<string> {
  return invoke<string>('skill_get_ssot_dir');
}

// ========== Security Scanning Types ==========

/**
 * Security finding severity level
 */
export type SecuritySeverity = 'critical' | 'high' | 'medium' | 'low' | 'info';

/**
 * Security finding category
 */
export type SecurityCategory =
  | 'command_execution'
  | 'code_injection'
  | 'filesystem_access'
  | 'network_access'
  | 'sensitive_data'
  | 'privilege_escalation'
  | 'obfuscated_code'
  | 'other';

/**
 * A single security finding from scanning
 */
export interface SecurityFinding {
  ruleId: string;
  title: string;
  description: string;
  severity: SecuritySeverity;
  category: SecurityCategory;
  filePath: string;
  line: number;
  column: number;
  snippet: string | null;
  suggestion: string | null;
}

/**
 * Summary statistics for a security scan
 */
export interface SecurityScanSummary {
  filesScanned: number;
  totalFindings: number;
  critical: number;
  high: number;
  medium: number;
  low: number;
  info: number;
  isSafe: boolean;
  riskScore: number;
}

/**
 * Complete security scan report
 */
export interface SecurityScanReport {
  skillId: string;
  skillName: string | null;
  scannedPath: string;
  scannedAt: number;
  durationMs: number;
  summary: SecurityScanSummary;
  findings: SecurityFinding[];
}

/**
 * Options for security scanning
 */
export interface SecurityScanOptions {
  maxFileSize?: number;
  maxFiles?: number;
  extensions?: string[];
  skipPatterns?: string[];
  minSeverity?: SecuritySeverity;
}

// ========== Security Scanning Commands ==========

/**
 * Scan an installed skill for security issues
 */
export async function scanInstalledSkill(
  directory: string,
  options?: SecurityScanOptions
): Promise<SecurityScanReport> {
  return invoke<SecurityScanReport>('skill_scan_installed', { directory, options });
}

/**
 * Scan a local path for security issues (pre-install check)
 */
export async function scanSkillPath(
  path: string,
  options?: SecurityScanOptions
): Promise<SecurityScanReport> {
  return invoke<SecurityScanReport>('skill_scan_path', { path, options });
}

/**
 * Get the number of security rules available
 */
export async function getSecurityRuleCount(): Promise<number> {
  return invoke<number>('skill_security_rule_count');
}

// ========== Utility Functions ==========

/**
 * Check if running in Tauri environment
 */
export function isNativeSkillAvailable(): boolean {
  return isTauri();
}

/**
 * Safe wrapper for skill operations that may fail in non-Tauri environments
 */
export async function safeSkillInvoke<T>(
  operation: () => Promise<T>,
  fallback: T
): Promise<T> {
  if (!isNativeSkillAvailable()) {
    return fallback;
  }
  try {
    return await operation();
  } catch (error) {
    console.error('Skill operation failed:', error);
    return fallback;
  }
}

File: src-tauri/src/commands/extensions/skill.rs
//! Skill Tauri Commands
//!
//! Exposes skill management functionality to the frontend.

use crate::skill::{
    AddRepoInput, DiscoverableSkill, InstallSkillInput, InstalledSkill, LocalSkill, Skill,
    SkillDiscoveryResult, SkillError, SkillRepo, SkillSearchFilters, SkillService, SkillState,
};
use std::path::PathBuf;
use std::sync::Arc;
use tauri::State;
use tokio::sync::RwLock;

/// Skill service state wrapper
pub struct SkillServiceState(pub Arc<RwLock<SkillService>>);

/// Create skill service for app state
pub fn create_skill_service(app_data_dir: PathBuf) -> SkillServiceState {
    match SkillService::new(app_data_dir) {
        Ok(service) => SkillServiceState(Arc::new(RwLock::new(service))),
        Err(e) => {
            log::error!("Failed to create skill service: {}", e);
            panic!("Failed to initialize skill service: {}", e);
        }
    }
}

// ========== Repository Commands ==========

/// List all configured skill repositories
#[tauri::command]
pub async fn skill_list_repos(
    state: State<'_, SkillServiceState>,
) -> Result<Vec<SkillRepo>, String> {
    let service = state.0.read().await;
    Ok(service.list_repos().await)
}

/// Add a skill repository
#[tauri::command]
pub async fn skill_add_repo(
    state: State<'_, SkillServiceState>,
    input: AddRepoInput,
) -> Result<(), String> {
    let service = state.0.read().await;
    let repo = SkillRepo {
        owner: input.owner,
        name: input.name,
        branch: input.branch.unwrap_or_else(|| "main".to_string()),
        enabled: true,
    };
    service.add_repo(repo).await.map_err(|e| e.to_string())
}

/// Remove a skill repository
#[tauri::command]
pub async fn skill_remove_repo(
    state: State<'_, SkillServiceState>,
    owner: String,
    name: String,
) -> Result<(), String> {
    let service = state.0.read().await;
    service
        .remove_repo(&owner, &name)
        .await
        .map_err(|e| e.to_string())
}

/// Toggle repository enabled state
#[tauri::command]
pub async fn skill_toggle_repo(
    state: State<'_, SkillServiceState>,
    owner: String,
    name: String,
    enabled: bool,
) -> Result<(), String> {
    let service = state.0.read().await;
    service
        .toggle_repo(&owner, &name, enabled)
        .await
        .map_err(|e| e.to_string())
}

// ========== Discovery Commands ==========

/// Discover skills from all enabled repositories
#[tauri::command]
pub async fn skill_discover(
    state: State<'_, SkillServiceState>,
) -> Result<Vec<DiscoverableSkill>, String> {
    let service = state.0.read().await;
    service.discover_skills().await.map_err(|e| e.to_string())
}

/// Get all skills (merged view)
#[tauri::command]
pub async fn skill_get_all(state: State<'_, SkillServiceState>) -> Result<Vec<Skill>, String> {
    let service = state.0.read().await;
    service.get_all_skills().await.map_err(|e| e.to_string())
}

/// Discover all skills (discoverable + installed + local)
#[tauri::command]
pub async fn skill_discover_all(
    state: State<'_, SkillServiceState>,
) -> Result<SkillDiscoveryResult, String> {
    let service = state.0.read().await;
    service.discover_all().await.map_err(|e| e.to_string())
}

/// Search and filter skills
#[tauri::command]
pub async fn skill_search(
    state: State<'_, SkillServiceState>,
    filters: SkillSearchFilters,
) -> Result<Vec<Skill>, String> {
    let service = state.0.read().await;
    service
        .search_skills(filters)
        .await
        .map_err(|e| e.to_string())
}

/// Scan for local unregistered skills
#[tauri::command]
pub async fn skill_scan_local(
    state: State<'_, SkillServiceState>,
) -> Result<Vec<LocalSkill>, String> {
    let service = state.0.read().await;
    service.scan_local_skills().await.map_err(|e| e.to_string())
}

// ========== Installation Commands ==========

/// Install a skill from repository
#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn skill_install(
    state: State<'_, SkillServiceState>,
    input: InstallSkillInput,
) -> Result<InstalledSkill, String> {
    let service = state.0.read().await;
    let repo_branch = input.branch.unwrap_or_else(|| "main".to_string());
    let key = format!("{}/{}:{}", input.owner, input.repo, input.directory);
    let skill = DiscoverableSkill {
        key,
        name: input.name.unwrap_or_else(|| input.directory.clone()),
        description: input.description.unwrap_or_default(),
        directory: input.directory,
        readme_url: input.readme_url,
        repo_owner: input.owner,
        repo_name: input.repo,
        repo_branch,
    };
    service
        .install_skill(&skill)
        .await
        .map_err(|e| e.to_string())
}

/// Install a skill from local path
#[tauri::command]
pub async fn skill_install_local(
    state: State<'_, SkillServiceState>,
    source_path: String,
    name: Option<String>,
) -> Result<InstalledSkill, String> {
    let service = state.0.read().await;
    let path = std::path::Path::new(&source_path);
    service
        .install_local_skill(path, name.as_deref())
        .await
        .map_err(|e| e.to_string())
}

/// Register a local skill already in SSOT directory
#[tauri::command]
pub async fn skill_register_local(
    state: State<'_, SkillServiceState>,
    directory: String,
) -> Result<InstalledSkill, String> {
    let service = state.0.read().await;
    service
        .register_local_skill(&directory)
        .await
        .map_err(|e| e.to_string())
}

/// Uninstall a skill
#[tauri::command]
pub async fn skill_uninstall(
    state: State<'_, SkillServiceState>,
    id: String,
) -> Result<(), String> {
    let service = state.0.read().await;
    service
        .uninstall_skill(&id)
        .await
        .map_err(|e| e.to_string())
}

// ========== State Management Commands ==========

/// Get all installed skills
#[tauri::command]
pub async fn skill_get_installed(
    state: State<'_, SkillServiceState>,
) -> Result<Vec<InstalledSkill>, String> {
    let service = state.0.read().await;
    Ok(service.get_installed_skills().await)
}

/// Get a specific installed skill
#[tauri::command]
pub async fn skill_get(
    state: State<'_, SkillServiceState>,
    id: String,
) -> Result<Option<InstalledSkill>, String> {
    let service = state.0.read().await;
    Ok(service.get_installed_skill(&id).await)
}

/// Get a specific installed skill or return error (uses SkillError)
#[tauri::command]
pub async fn skill_get_required(
    state: State<'_, SkillServiceState>,
    id: String,
) -> Result<InstalledSkill, String> {
    let service = state.0.read().await;
    service
        .get_skill_or_error(&id)
        .await
        .map_err(|e: SkillError| e.to_string())
}

/// Get skill state (legacy compatibility)
#[tauri::command]
pub async fn skill_get_state(
    state: State<'_, SkillServiceState>,
    id: String,
) -> Result<SkillState, String> {
    let service = state.0.read().await;
    service
        .get_skill_state(&id)
        .await
        .ok_or_else(|| "Skill not found".to_string())
}

/// Validate if a skill can be installed
#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn skill_validate_install(
    state: State<'_, SkillServiceState>,
    input: InstallSkillInput,
) -> Result<(), String> {
    let service = state.0.read().await;
    let repo_branch = input.branch.unwrap_or_else(|| "main".to_string());
    let key = format!("{}/{}:{}", input.owner, input.repo, input.directory);
    let skill = DiscoverableSkill {
        key,
        name: input.name.unwrap_or_else(|| input.directory.clone()),
        description: input.description.unwrap_or_default(),
        directory: input.directory,
        readme_url: input.readme_url,
        repo_owner: input.owner,
        repo_name: input.repo,
        repo_branch,
    };
    service
        .validate_install(&skill)
        .await
        .map_err(|e: SkillError| e.to_string())
}

/// Enable a skill
#[tauri::command]
pub async fn skill_enable(state: State<'_, SkillServiceState>, id: String) -> Result<(), String> {
    let service = state.0.read().await;
    service.enable_skill(&id).await.map_err(|e| e.to_string())
}

/// Disable a skill
#[tauri::command]
pub async fn skill_disable(state: State<'_, SkillServiceState>, id: String) -> Result<(), String> {
    let service = state.0.read().await;
    service.disable_skill(&id).await.map_err(|e| e.to_string())
}

/// Update skill metadata
#[tauri::command]
pub async fn skill_update(
    state: State<'_, SkillServiceState>,
    id: String,
    category: Option<String>,
    tags: Option<Vec<String>>,
) -> Result<(), String> {
    let service = state.0.read().await;
    service
        .update_skill(&id, category, tags)
        .await
        .map_err(|e| e.to_string())
}

// ========== Content Commands ==========

/// Read skill content (SKILL.md)
#[tauri::command]
pub async fn skill_read_content(
    state: State<'_, SkillServiceState>,
    directory: String,
) -> Result<String, String> {
    let service = state.0.read().await;
    service
        .read_skill_content(&directory)
        .map_err(|e| e.to_string())
}

/// Write skill content (SKILL.md)
#[tauri::command]
pub async fn skill_write_content(
    state: State<'_, SkillServiceState>,
    directory: String,
    content: String,
) -> Result<(), String> {
    let service = state.0.read().await;
    service
        .write_skill_content(&directory, &content)
        .map_err(|e| e.to_string())
}

/// List skill resource files
#[tauri::command]
pub async fn skill_list_resources(
    state: State<'_, SkillServiceState>,
    directory: String,
) -> Result<Vec<String>, String> {
    let service = state.0.read().await;
    service
        .list_skill_resources(&directory)
        .map_err(|e| e.to_string())
}

/// Read a skill resource file
#[tauri::command]
pub async fn skill_read_resource(
    state: State<'_, SkillServiceState>,
    directory: String,
    resource_path: String,
) -> Result<String, String> {
    let service = state.0.read().await;
    service
        .read_skill_resource(&directory, &resource_path)
        .map_err(|e| e.to_string())
}

/// Write a skill resource file
#[tauri::command]
pub async fn skill_write_resource(
    state: State<'_, SkillServiceState>,
    directory: String,
    resource_path: String,
    content: String,
) -> Result<(), String> {
    let service = state.0.read().await;
    service
        .write_skill_resource(&directory, &resource_path, &content)
        .map_err(|e| e.to_string())
}

/// Get SSOT directory path
#[tauri::command]
pub async fn skill_get_ssot_dir(state: State<'_, SkillServiceState>) -> Result<String, String> {
    let service = state.0.read().await;
    Ok(service.get_ssot_dir().to_string_lossy().to_string())
}

// ========== Security Scanning Commands ==========

use crate::skill::{SecurityScanOptions, SecurityScanReport, SkillSecurityScanner};

/// Scan an installed skill for security issues
#[tauri::command]
pub async fn skill_scan_installed(
    state: State<'_, SkillServiceState>,
    directory: String,
    options: Option<SecurityScanOptions>,
) -> Result<SecurityScanReport, String> {
    let service = state.0.read().await;
    let scanner = SkillSecurityScanner::new();
    let opts = options.unwrap_or_default();
    scanner
        .scan_installed(service.get_ssot_dir(), &directory, &opts)
        .map_err(|e| e.to_string())
}

/// Scan a local path for security issues (pre-install check)
#[tauri::command]
pub async fn skill_scan_path(
    path: String,
    options: Option<SecurityScanOptions>,
) -> Result<SecurityScanReport, String> {
    let scanner = SkillSecurityScanner::new();
    let opts = options.unwrap_or_default();
    let skill_path = std::path::Path::new(&path);
    if !skill_path.exists() {
        return Err(format!("Path does not exist: {}", path));
    }
    scanner.scan(skill_path, &opts).map_err(|e| e.to_string())
}

/// Get the number of security rules available
#[tauri::command]
pub fn skill_security_rule_count() -> usize {
    SkillSecurityScanner::new().rule_count()
}
